File: Bisection/Bisection.py
import timeit

fx = lambda x: x**2 - 11

def bisection_method(f,a, b, tol):
    if f(a)*f(b) > 0:
        print("No root found.")
    else:
        while (b - a) >= tol:
            midpoint = (a + b)/2.0
            if f(midpoint) == 0:
                return(midpoint) 
            elif f(a)*f(midpoint) < 0: 
                b = midpoint
            else:
                a = midpoint
        return(midpoint)
    
sample=10000
execution_time = timeit.timeit(lambda:bisection_method(fx,-1, 5, 0.0001), number=sample)

print("Average Time:", execution_time/sample)
print("Result:",bisection_method(fx,-1, 5, 0.0001))

File: Bisection/Bisection.jl
using BenchmarkTools

f(x) =  x^2 - 11
function bisection_method(f,a, b, tol)
    if f(a)*f(b) > 0
        println("No root found.")
    else
        while (b - a) >= tol
            midpoint = (a + b) / 2.0
            if f(midpoint) == 0
                return midpoint
            elseif f(a) * f(midpoint) < 0
                b = midpoint
            else
                a = midpoint
            end
        end       
        return (a + b) / 2.0
    end
end

result = @benchmark bisection_method(f,-1, 5, 0.0001)

println("Answer: ", bisection_method(f,-1, 5, 0.0001))
println("Minimum time: $(minimum(result.times) / 1e9) seconds")
println("Maximum time: $(maximum(result.times) / 1e9) seconds")
println("Average time: $(mean(result.times) / 1e9) seconds")
